sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1025,7 +1025,8 @@ object PushDownPredicate extends Rule[LogicalPlan] with PredicateHelper {
     // state and all the input rows processed before. In another word, the order of input rows
     // matters for non-deterministic expressions, while pushing down predicates changes the order.
     case filter @ Filter(condition, project @ Project(fields, grandChild))
-      if fields.forall(_.deterministic) =>
+        if fields.forall(_.deterministic) &&
+          fields.forall(_.find(_.isInstanceOf[ScalaUDF]).isEmpty) =>
```

Comment:
Can you look into it to see where we made this assumption? For now I prefer to override `deterministic` in `ScalaUDF`, if it's not a lot of effort to fix it.